File: unity/Assets/Script/Scene/TeamRoom.cs
/**
 * Copyright 2023. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */



using Com.Huawei.Game.Gobes;
using System.Collections;
using Com.Huawei.Game.Gobes.Store;
using Com.Huawei.Game.Gobes.Utils;
using UnityEngine;
using UnityEngine.UI;

public class TeamRoom : MonoBehaviour
{

    public Text PlayerOwnerName = null;

    public Text PlayerOneName = null;

    public Text PlayerTwoName = null;

    public Text PlayerThreeName = null;

    public Text PlayerOneStatus = null;

    public Text PlayerTwoStatus = null;

    public Text PlayerThreeStatus = null;

    public Button UnReadyBtnOne = null;

    public Button UnReadyBtnTwo = null;

    public Button UnReadyBtnThree = null;

    public Button ReadyBtnOne = null;

    public Button ReadyBtnTwo = null;

    public Button ReadyBtnThree = null;

    public Button StartBtn = null;

    public Button DisableStartBtn = null;

    public Dialog Dialog = null;

    public Reloading Loading = null;

    public static readonly int FontSize = 14;

    
    // -1 断线中 0 断线重连 1 重连成功 2 重连失败 
    private FrameSync.ReConnectState isReConnect = FrameSync.ReConnectState.reConnectionDefault;

    // 定时
    private float interval = 2f; // 每隔2秒执行一次
    private float count = 0;

    // Start is called before the first frame update
    void Start()
    {
        InitRoomView();
        InitListener();
        if(Global.Room.roomInfo.RoomStatus ==(int)RoomStatus.SYNCING ) {
            OnStartFrameSync();
        }
    }

    // ====================刷新房间信息====================
    // Update is called once per frame
    void Update()
    {
        // 定时任务
        count += Time.deltaTime;
        if (count >= interval)
        {
            count = 0;
            InitRoomView();
            if (isReConnect == FrameSync.ReConnectState.reConnectIng)
            {
                ReConnectRoom();
            }
        }
    }

    public void InitRoomView()
    {
        if (Global.Room != null)
        {
            Global.Room.Update(response =>
            {
                if (response.RtnCode == 0)
                {
                    UnityMainThread.wkr.AddJob(SetRoomView);
                }
            });
        }
    }

    void InitListener() {
        Util.SaveRoomType(FrameSync.RoomType.TEAMROOM);
        // 监听加入房间
        Global.Room.OnJoin = playerInfo => OnJoining();
        // 监听开始帧同步
        Global.Room.OnStartSyncFrame = () => OnStartFrameSync();
        // 断线重连
        Global.Room.OnDisconnect = playerInfo => OnDisconnect(playerInfo); // 断连监听
        // 离开房间
        Global.room.OnLeave = playerInfo => OnLeaving(playerInfo);
    }

    public void SetRoomView() {
        InitAllBtnActive();
        RoomInfo roomInfo = Global.Room != null ? Global.Room.roomInfo : null;
        Global.player = Global.Room._player;
        bool allPlayersStatus = true;
        if (roomInfo.Players.Length > 0) {
            string redTeam = "0";
            ArrayList players = new ArrayList();

            int playerNo = 1;

            for (int i = 0; i < roomInfo.Players.Length; i++) {
                PlayerInfo player = roomInfo.Players[i];
                if (player.IsRobot == 1)
                {
                    player.CustomPlayerStatus = 1;
                    player.CustomPlayerProperties = player.RobotName;
                }

                // 渲染在线玩家
  
                    // 非房主
                    if (player.PlayerId != roomInfo.OwnerId)
                    {
                        players.Add(player);
                        if (player.CustomPlayerStatus == 0)
                        {
                            allPlayersStatus = false;
                        }
                    }
                    else {
                        // 房主
                        PlayerOwnerName.text = player.CustomPlayerProperties != null ? player.CustomPlayerProperties : player.PlayerId;
                    }
                
                if (player.PlayerId == roomInfo.OwnerId)
                {
                    redTeam = player.TeamId;
                }

            }
            for (int i = 0; i < players.Count; i++)
            {
                PlayerInfo player = (PlayerInfo)players[i];
                if (player.TeamId == redTeam)
                {
                    SetRedPlayer(player);
                }
                else
                {
                    SetYellowPlayer(player, playerNo);
                    playerNo = 3;
                }
            }
        }
        if (roomInfo.OwnerId != Global.playerId)
        {
            StartBtn.gameObject.SetActive(false);
        }
        else
        {
            StartBtn.gameObject.SetActive(true);
            StartBtn.interactable = allPlayersStatus;
        }
    }

    public void SetYellowPlayer(PlayerInfo player, int playerNo)
    {
        bool isPlayerStatus = player.CustomPlayerStatus == 1 ? true :false;
        string playerStatus = player.CustomPlayerProperties;

        switch (playerNo)
        {
            case 1:

                if (player.PlayerId == Global.playerId) {
                    UnReadyBtnOne.gameObject.SetActive(isPlayerStatus);
                    ReadyBtnOne.gameObject.SetActive(!isPlayerStatus);     
                }
                PlayerOneStatus.text = isPlayerStatus ? "已准备" : "未准备";
                if (player.IsRobot==1)
                {
                    PlayerOneName.fontSize = FontSize;
                }
                PlayerOneName.text = player.CustomPlayerProperties != null ? player.CustomPlayerProperties : player.PlayerId;
                break;
            case 3:
                if (player.PlayerId == Global.playerId) {
                    UnReadyBtnThree.gameObject.SetActive(isPlayerStatus);
                    ReadyBtnThree.gameObject.SetActive(!isPlayerStatus);
                }
                PlayerThreeStatus.text = isPlayerStatus ? "已准备" : "未准备";
                if (player.IsRobot==1)
                {
                    PlayerThreeName.fontSize = FontSize;
                }
                PlayerThreeName.text = player.CustomPlayerProperties != null ? player.CustomPlayerProperties : player.PlayerId;
                break;
        }
    }

    public void SetRedPlayer(PlayerInfo player) {
        bool isPlayerStatus = player.CustomPlayerStatus == 1 ? true : false;
        if (player.PlayerId == Global.playerId)
        {
            UnReadyBtnTwo.gameObject.SetActive(isPlayerStatus);
            ReadyBtnTwo.gameObject.SetActive(!isPlayerStatus);
        }
        PlayerTwoStatus.text = isPlayerStatus ? "已准备" : "未准备";
        if (player.IsRobot==1)
        {
            PlayerTwoName.fontSize = FontSize;
        }
        PlayerTwoName.text = player.CustomPlayerProperties != null ? player.CustomPlayerProperties : player.PlayerId;
    }

    void ReConnectRoom() {
        // 没有超过重连时间，就进行重连操作
        try {
            // 没有超过重连时间，就进行重连操作
            Global.room.Reconnect(response => {
                if (response.RtnCode == 0) {
                    // 重连成功
                    Debug.Log("玩家重连成功");
                    UnityMainThread.wkr.AddJob(() => {
                        GameObject loading = GameObject.Find("/loading2(Clone)");
                        Destroy(loading);
                        isReConnect = FrameSync.ReConnectState.reConnectionDefault;
                    });
                }
                if (response.RtnCode == ((int)ErrorCode.SDK_NOT_IN_ROOM))
                {
                    Debug.Log("玩家重连失败");
                    UnityMainThread.wkr.AddJob(() => {
                        Route.GoHall();
                        isReConnect = FrameSync.ReConnectState.reConnectionDefault;
                    });
                }
            });
        } catch (SDKException e){
            SDKDebugLogger.Log(e.Message);
            if (e.code == (int) ErrorCode.INVALID_ROOM || e.code == (int) ErrorCode.PLAYERS_EXCEEDS_ROOM_MAX 
                || e.code == (int) ErrorCode.INVALID_ROOM_STATUS) {
                // 重连失败
                Debug.Log("重连失败");
                UnityMainThread.wkr.AddJob(() => {
                    Route.GoHall();
                    isReConnect = FrameSync.ReConnectState.reConnectionDefault;
                });
            } else {
                SDKDebugLogger.Log("玩家持续重连中...");
            }
        }
    }

    private void InitAllBtnActive() {
        PlayerOneStatus.text = "";
        PlayerTwoStatus.text = "";
        PlayerThreeStatus.text = "";
        ReadyBtnOne.gameObject.SetActive(false);
        ReadyBtnTwo.gameObject.SetActive(false);
        ReadyBtnThree.gameObject.SetActive(false);
        UnReadyBtnOne.gameObject.SetActive(false);
        UnReadyBtnTwo.gameObject.SetActive(false);
        UnReadyBtnThree.gameObject.SetActive(false);
        StartBtn.gameObject.SetActive(false);
        PlayerOwnerName.text = "";
        PlayerOneName.text = "";
        PlayerTwoName.text = "";
        PlayerThreeName.text = "";
    }

    // 准备
    public void Ready()
    {
        Debug.Log("准备");
        int ready = 1;
        Global.player.UpdateCustomStatus(ready, response =>
        {
            if (response.RtnCode == 0) {
                // 修改玩家自定义状态
                InitRoomView();
            } else {
                // 修改玩家自定义状态失败
                Dialog.Open("提示", "准备就绪失败" + Util.ErrorMessage(response));
            }
        });
    }

    // 取消准备
    public void CancelReady(){
        Debug.Log("取消准备");
        int unready = 0;
        Global.player.UpdateCustomStatus(unready, response =>
        {
            if (response.RtnCode == 0)
            {
                // 修改玩家自定义状态
                InitRoomView();
            } else {
                // 修改玩家自定义状态失败
                Dialog.Open("提示", "取消准备失败" + Util.ErrorMessage(response));
            }
        });
    }

    // 开始组队匹配游戏
    public void StartTeamGame()
    {
        Debug.Log("开始组队匹配游戏");
        Global.Room.Update(response => {
            if (response.RtnCode == 0)
            {
                int readyStatus = 1;
                PlayerInfo[] players = response.RoomInfo.Players;
                foreach (PlayerInfo player in players)
                {
                    if (player.IsRobot == 1)
                    {
                        player.CustomPlayerStatus = 1;
                    }

                    if (player.PlayerId != response.RoomInfo.OwnerId)
                    {
                        if (player.CustomPlayerStatus == 0)
                        {
                            readyStatus = 0;
                            return;
                        }
                    }
                }
                if (readyStatus == 0)
                {
                    Dialog.Open("提示", "还有玩家未准备，请稍后！");
                }
                else
                {
                    Global.room.StartFrameSync(res => {
                        if (res.RtnCode == 0)
                        {
                            // 开始帧同步成功
                            Debug.Log("开始帧同步成功");
                            UnityMainThread.wkr.AddJob(() => {
                                OnStartFrameSync();
                            });
                        }
                        else
                        {
                            // 开始帧同步失败
                            Dialog.Open("提示", "开始帧同步失败" + Util.ErrorMessage(res));
                        }
                    });
                }
            }
        });
    }
    

    // ====================广播====================
    void OnJoining()
    {
        InitRoomView();
    }

    void OnLeaving(FramePlayerInfo playerInfo)
    {
        Debug.Log("广播--离开房间");
        if (Global.playerId == playerInfo.PlayerId)
        {
            UnityMainThread.wkr.AddJob(Route.GoHall);
        }
        else
        {
            InitRoomView();
        }
    }

    void OnDisconnect(FramePlayerInfo playerInfo) {
        Debug.Log("广播--组队玩家掉线");
        if (playerInfo.PlayerId == Global.playerId) {
            UnityMainThread.wkr.AddJob(() => {
                Reloading loading = Instantiate(Loading);
                loading.Open("重连中...");
                isReConnect = FrameSync.ReConnectState.reConnectIng;
            });
        }
    }

    void OnStartFrameSync() {
        Debug.Log("开始帧同步");
        Global.state = 1;
        Global.keyOperate = 1;
        UnityMainThread.wkr.AddJob(Route.GoGameView);
    }

}

File: cocos/assets/Script/scene/teamroom.ts
/**
 * Copyright 2024. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import * as Util from "../../util";
import {RoomType} from "../commonValue";
import global from "../../global";
import Dialog from "../comp/Dialog";
import {PlayerInfo} from "../../GOBE/GOBE";
import {setRoomType, sleep} from "../function/Common";
import {GameSceneType} from "../function/FrameSync";


export enum PlayerOnline {
    online = 1,  // 在线
    offline = 3,  // 离线
}

export enum Player {
    one = 1,  // 玩家1
    two = 2,  // 玩家2
    three = 3,// 玩家3
}

const {ccclass, property} = cc._decorator;

@ccclass
export default class TeamRoom extends cc.Component {

    @property(cc.Label)
    ownerName: cc.Label = null;

    @property(cc.Label)
    playerOneName: cc.Label = null;

    @property(cc.Label)
    playerTwoName: cc.Label = null;

    @property(cc.Label)
    playerThreeName: cc.Label = null;

    @property(cc.Node)
    unReadyOneBtn: cc.Node = null;

    @property(cc.Node)
    readyOneBtn: cc.Node = null;

    @property(cc.Node)
    unReadyTwoBtn: cc.Node = null;

    @property(cc.Node)
    readyTwoBtn: cc.Node = null;

    @property(cc.Node)
    unReadyThreeBtn: cc.Node = null;

    @property(cc.Node)
    readyThreeBtn: cc.Node = null;

    @property(cc.Node)
    disableBtnStart: cc.Node = null;

    @property(cc.Node)
    enableBtnStart: cc.Node = null;

    @property(cc.Label)
    playerOneStatus: cc.Label = null;

    @property(cc.Label)
    playerTwoStatus: cc.Label = null;

    @property(cc.Label)
    playerThreeStatus: cc.Label = null;

    @property(cc.Prefab)
    dialogPrefab: cc.Prefab = null;

    start() {
        this.initView();
        this.initListener();
        this.initSchedule();
        if(global.room.isSyncing){
            global.gameSceneType = GameSceneType.FOR_GAME;
            cc.director.loadScene("game");
        }
    }

    initView() {
        this.initDialog();
        this.initRoomView();
    }

    initDialog() {
        // 设置对话框
        const dialogNode = cc.instantiate(this.dialogPrefab) as cc.Node;
        dialogNode.parent = this.node;
    }

    initRoomView() {
        if (global.room) {
            // 偶尔网络波动，不跳转到大厅，不然其他用户的显示会有问题
            global.room.update().then(() => this.setRoomView());
        }
    }

    setRoomView() {
        this.initAllBtnActive();
        const roomInfo = global.room;
        let allPlayersStatus = true;
        if (roomInfo.players.length > 0) {
            let redTeam = "0";
            let players = [];
            let playerNo = 1;
            for (let i = 0; i < roomInfo.players.length; i++) {
                let player = roomInfo.players[i];
                if (player.status === PlayerOnline.online || player.isRobot === 1) {
                    // 渲染在线玩家
                    if (player.playerId !== roomInfo.ownerId) {
                        // 非房主
                        players.push(player);
                        if (player.customPlayerStatus === 0) {
                            allPlayersStatus = false;
                        }
                    } else {
                        // 房主
                        let playerProperties = JSON.parse(player.customPlayerProperties);
                        this.ownerName.string = playerProperties["playerName"] ?
                            playerProperties["playerName"] : player.playerId;
                    }
                }
                if (player.playerId === roomInfo.ownerId) {
                    redTeam = player.teamId;
                }
            }
            for (let i = 0; i < players.length; i++) {
                let player = players[i];
                if (player.teamId === redTeam) {
                    this.setRedPlayer(player);
                } else {
                    this.setYellowPlayer(player, playerNo);
                    playerNo = 3;
                }
            }
        }
        if (roomInfo.ownerId !== global.playerId) {
            this.disableBtnStart.active = false;
            this.enableBtnStart.active = false;
        } else {
            this.disableBtnStart.active = !allPlayersStatus;
            this.enableBtnStart.active = allPlayersStatus;
        }
    }

    private setYellowPlayer(player, playerNo: number) {
        let isPlayerStatus = player.customPlayerStatus === 1;
        let playerName: string;

        if (player.isRobot === 1) {
            playerName = player.robotName || `机器人${player.playerId}`;
            switch (playerNo) {
                case 1:
                    this.playerOneName.fontSize = 10;
                    this.playerOneStatus.string = "已准备";
                    this.playerOneName.string = playerName;
                    this.unReadyOneBtn.active = false;
                    this.readyOneBtn.active = false;
                    break;
                case 3:
                    this.playerThreeName.fontSize = 10;
                    this.playerThreeStatus.string = "已准备";
                    this.playerThreeName.string = playerName;
                    this.unReadyThreeBtn.active = false;
                    this.readyThreeBtn.active = false;
                    break;
            }
        } else {
            let playerProperties = JSON.parse(player.customPlayerProperties);
            playerName = playerProperties["playerName"];
            switch (playerNo) {
                case 1:
                    if (player.playerId === global.playerId) {
                        this.unReadyOneBtn.active = isPlayerStatus;
                        this.readyOneBtn.active = !(isPlayerStatus);
                    }
                    this.playerOneStatus.string = isPlayerStatus ? "已准备" : "未准备";
                    this.playerOneName.string = playerName ?
                        playerName : player.playerId;
                    break;
                case 3:
                    if (player.playerId === global.playerId) {
                        this.unReadyThreeBtn.active = isPlayerStatus;
                        this.readyThreeBtn.active = !(isPlayerStatus);
                    }
                    this.playerThreeStatus.string = isPlayerStatus ? "已准备" : "未准备";
                    this.playerThreeName.string = playerName ?
                        playerName : player.playerId;
                    break;
            }
        }
    }

    private setRedPlayer(player) {
        if (player.isRobot === 1) {
            this.playerTwoName.fontSize = 10;
            this.playerTwoStatus.string = "已准备";
            this.playerTwoName.string = player.robotName || `机器人${player.playerId}`;
            this.unReadyTwoBtn.active = false;
            this.readyTwoBtn.active = false;
        } else {
            let isPlayerStatus = player.customPlayerStatus === 1;
            if (player.playerId === global.playerId) {
                this.unReadyTwoBtn.active = isPlayerStatus;
                this.readyTwoBtn.active = !(isPlayerStatus);
            }
            this.playerTwoStatus.string = isPlayerStatus ? "已准备" : "未准备";
            let playerProperties = JSON.parse(player.customPlayerProperties);
            this.playerTwoName.string = playerProperties["playerName"] ?
                playerProperties["playerName"] : player.playerId;
        }

    }

    initListener() {
        this.unReadyOneBtn.on(cc.Node.EventType.TOUCH_START, () => this.cancelReady());
        this.readyOneBtn.on(cc.Node.EventType.TOUCH_START, () => this.ready());
        this.unReadyTwoBtn.on(cc.Node.EventType.TOUCH_START, () => this.cancelReady());
        this.readyTwoBtn.on(cc.Node.EventType.TOUCH_START, () => this.ready());
        this.unReadyThreeBtn.on(cc.Node.EventType.TOUCH_START, () => this.cancelReady());
        this.readyThreeBtn.on(cc.Node.EventType.TOUCH_START, () => this.ready());
        this.enableBtnStart.on(cc.Node.EventType.TOUCH_START, () => this.startTeamGame());

        // 监听加入房间
        global.room.onJoin(() => this.onJoining());
        // 监听开始帧同步
        global.room.onStartFrameSync(() => this.onStartFrameSync());
        // 断线重连
        global.room.onDisconnect((playerInfo: PlayerInfo) => this.onDisconnect(playerInfo)); // 断连监听
        // 离开房间
        global.room.onLeave((playerInfo) => this.onLeaving(playerInfo))
    }

    async onDisconnect(playerInfo: PlayerInfo) {
        Util.printLog("玩家掉线");
        if (playerInfo.playerId === global.playerId) {
            global.isConnected = false;
            this.reConnectRoom();
        }
    }

    async reConnectRoom() {
        // 没有超过重连时间，就进行重连操作
        while (!global.isConnected){
            await global.room.reconnect().then(() => {
                global.isConnected = true;
                Util.printLog("玩家重连房间成功");
            }).catch((error) => {
                if (!error.code || error.code === 91002) {
                    // 加入房间请求不通就继续重连
                    Util.printLog("玩家重连房间失败，重新尝试");
                }
                if (error.code && (error.code === 101117 || error.code === 101107 || error.code === 101120)) {
                    // 无法加入房间需要退出到大厅
                    Util.printLog("玩家重连房间失败");
                    cc.director.loadScene("hall");
                }
            });
            await sleep(2000).then();
        }
    }

    leaveRoom() {
        Util.printLog(`正在退出房间`);
        global.client.leaveRoom().then((client) => {
            // 退出房间成功
            global.roomType = RoomType.NULL;
            global.client = client;
            Util.printLog("退出房间成功");
        }).catch((e) => {
            // 退出房间失败
            Dialog.open("提示", "退出房间失败" + Util.errorMessage(e));
        });
    }

    initAllBtnActive() {
        this.unReadyOneBtn.active = false;
        this.readyOneBtn.active = false;
        this.playerOneStatus.string = "";
        this.unReadyTwoBtn.active = false;
        this.readyTwoBtn.active = false;
        this.playerTwoStatus.string = "";
        this.unReadyThreeBtn.active = false;
        this.readyThreeBtn.active = false;
        this.playerThreeStatus.string = "";

        this.ownerName.string = "";
        this.playerOneName.string = "";
        this.playerTwoName.string = "";
        this.playerThreeName.string = "";

        this.disableBtnStart.active = false;
        this.enableBtnStart.active = false;
    }

    // 准备
    ready() {
        Util.printLog("准备");
        let ready = 1;
        global.player.updateCustomStatus(ready);
        // 修改玩家自定义状态
        this.initRoomView();
    }

    // 取消准备
    cancelReady() {
        Util.printLog("取消准备");
        let unready = 0;
        global.player.updateCustomStatus(unready);
        // 修改玩家自定义状态
        this.initRoomView();
    }

    // 开始组队匹配游戏
    startTeamGame() {
        Util.printLog(`开始组队匹配游戏`);
        global.room.update().then((room) => {
            let readyStatus = 1;
            room.players.forEach(function (player) {
                if (player.playerId != room.ownerId) {
                    // @ts-ignore
                    if (player.customPlayerStatus === 0) {
                        readyStatus = 0;
                        return;
                    }
                }
            });
            if (readyStatus === 0) {
                Dialog.open("提示", "还有玩家未准备，请稍后！");
            } else {
                global.room.startFrameSync().then(() => {
                    // 开始帧同步成功
                    Util.printLog("开始帧同步成功");
                }).catch((e) => {
                    // 开始帧同步失败
                    Dialog.open("提示", "开始帧同步失败" + Util.errorMessage(e));
                });
            }
        });
    }

    onDisable() {
        // 关闭对话框
        Dialog.close();
        if (global.room) {
            global.room.removeAllListeners();
        }
    }

    // ====================定时去刷新房间信息====================
    initSchedule() {
        // 以秒为单位的时间间隔
        let interval = 2;
        // 开始延时
        let delay = 5;
        this.schedule(function () {
            this.initRoomView();
        }, interval, cc.macro.REPEAT_FOREVER, delay);
    }

    // ====================广播====================
    onJoining() {
        setRoomType(RoomType.TwoVTwo);
        this.initRoomView()
    }

    onLeaving(playerInfo: PlayerInfo) {
        Util.printLog("广播--离开房间");
        if (global.playerId === playerInfo.playerId) {
            cc.director.loadScene("hall");
        } else {
            this.initRoomView()
        }
    }

    onStartFrameSync() {
        Util.printLog("广播--开始帧同步");
        global.gameSceneType = GameSceneType.FOR_GAME;
        cc.director.loadScene("game");
    }

}
